aws_lambda_builders/workflows/nodejs_npm/DESIGN.md

```diff
@@ -0,0 +1,126 @@
+## NodeJS - NPM Lambda Builder
+
+### Scope
+
+This package is an effort to port the Claudia.JS packager to a library that can
+be used to handle the dependency resolution portion of packaging NodeJS code
+for use in AWS Lambda. The scope for this builder is to take an existing
+directory containing customer code, including a valid `package.json` manifest
+specifying third-party dependencies. The builder will use NPM to include
+production dependencies and exclude test resources in a way that makes them
+deployable to AWS Lambda.
+
+### Challenges
+
+NPM normally stores all dependencies in a `node_modules` subdirectory. It
+supports several dependency categories, such as development dependencies
+(usually third-party build utilities and test resources), optional dependencies
+(usually required for local execution but already available on the production
+environment, or peer-dependencies for optional third-party packages) and
+production dependencies (normally the minimum required for correct execution).
+All these dependency types are mixed in the same directory.
+
+To speed up Lambda startup time and optimise usage costs, the correct thing to
+do in most cases is just to package up production dependencies. During development 
+work we can expect that the local `node_modules` directory contains all the 
+various dependency types, and NPM does not provide a way to directly identify
+just the ones relevant for production. To identify production dependencies, 
+this packager needs to copy the source to a clean temporary directory and re-run
+dependency installation there.
+  
+It's reasonable to expect that some developers will not carefully separate
```

Comment:
I guess the option Sanath and I discussed is more to do with optional dependencies than dev dependencies, so I'll clarify that in the design doc. 

NPM has the concept of optional dependencies, very frequently used for excluding aws-sdk for from the deployment package, but leaving it as a non-dev dependency for small lambdas. AWS SDK is already available on the Lambda VM, so this makes deployment significantly faster for single-file lambdas, for example. I dislike this because it's not 100% clear what version of the API gets used, but people do this frequently enough that we should handle it in some way. the proposal was to use an env var to override NPM arguments.

Reply:
I've updated this in https://github.com/awslabs/aws-lambda-builders/pull/44/commits/fdc2707266b97add790aa7b1a653bfef832685bf

Reply:
That makes sense. Thanks for clarifing it.